core/mavlink_system.cpp

```diff
@@ -136,6 +137,16 @@ void MAVLinkSystem::process_mavlink_message(const mavlink_message_t &message)
             it->callback(message);
             _mavlink_handler_table_mutex.lock();
         }
+
+        if (_iterator_invalidated) {
+            // Someone messed with the map while we were doing the callback.
```

Comment:
How can that happen? Don't we have the `_mavlink_handler_table_mutex` lock for preventing that?

Reply:
The problem is that we have to release the lock during the callback. If we didn't release the lock during the callback, we could lock ourselves out if the code in the callback tries to register or unregister a message.

An alternative would be to use a recursive_lock, so we wouldn't lock ourselves out, however, we would still change the map and invalidate the iterator.

Reply:
The proper solution is to have events in queues instead of callbacks, I think but we're far away from that :disappointed:.

Reply:
I see. Well, that's fine for now :-).